File: RainWaterTrapping.js
const rainWaterTrapping=(arr,n)=>{
    let left=mal(arr,n)
    let right = mar(arr,n)
    console.log(left,right)
    let water=0
    for(let i=0;i<n;i++){
        water+=Math.min(left[i],right[i])-arr[i]
    }
    return water;
}
const mal=(arr,n)=>{
    let left=[]
    for(let i=0;i<n;i++){
        if(i==0){
            left.push(arr[0])
        }
        else if(left[i-1]>arr[i]){
            left.push(left[i-1])
        }else{
            left.push(arr[i])
        }
    }
    return left
}

const mar=(arr,n)=>{
    let right=[]
    for(let i=n-1;i>=0;i--){
        if(i==n-1){
            right.push(arr[n-1])
        }
        else if(right[n-2-i]>arr[i]){
            right.push(right[n-2-i])
        }else{
            right.push(arr[i])
        }
    }
    return right.reverse()
}
console.log(rainWaterTrapping([3,0,5,2,0,4],6))

File: Python Problems/RainWaterTrapping.py

from operator import le
from typing import List


class Solution:
    def trap(self, height: List[int]) -> int:
        left=[]
        for i in range(len(height)):
            if i==0:
                left.append(height[i])
            elif left[i-1]>height[i]:
                left.append(left[i-1])
            else:
                left.append(height[i])
        rigth=[]
        for i in range(len(height)-1,-1,-1):
            if i==len(height)-1:
                rigth.append(height[i])
            elif rigth[len(height)-2-i]>height[i]:
                rigth.append(rigth[len(height)-2-i])
            else:
                rigth.append(height[i])
        rigth.reverse()
        water=0
        for i in range(len(height)):
            water=water+ (min(left[i],rigth[i])-height[i])
        return water


ans= Solution()
print(ans.trap([3,0,0,2,0,4]))